pkg/query-service/app/preferences/model.go

```diff
@@ -0,0 +1,510 @@
+
+	err := preference.IsValidValue(preferenceValue)
+	if err != nil {
+		return nil, err
+	}
+
+	// update the values in the org_preference table and return the key and the value
+	query := `INSERT INTO org_preference(preference_id,preference_value,org_id) VALUES($1,$2,$3)
+	ON CONFLICT(preference_id,org_id) DO
+	UPDATE SET preference_value= $2 WHERE preference_id=$1 AND org_id=$3;`
+
+	_, dberr := db.Exec(query, preferenceId, preferenceValue, orgId)
+
+	if dberr != nil {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in setting the preference value: %s", dberr.Error())}
+	}
+
+	return &PreferenceKV{
+		PreferenceId:    preferenceId,
+		PreferenceValue: preferenceValue,
+	}, nil
+}
+
+func GetAllOrgPreferences(ctx context.Context, orgId string) (*[]AllPreferences, *model.ApiError) {
+	// filter out all the org enabled preferences from the preference variable
+	allOrgPreferences := []AllPreferences{}
+
+	// fetch all the org preference values stored in org_preference table
+	orgPreferenceValues := []PreferenceKV{}
+
+	query := `SELECT preference_id,preference_value FROM org_preference WHERE org_id=$1;`
+	err := db.Select(&orgPreferenceValues, query, orgId)
+
+	if err != nil {
+		return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("error in getting all org preference values: %s", err)}
+	}
+
+	// create a map of key vs values from the above response
+	preferenceValueMap := map[string]interface{}{}
+
+	for _, preferenceValue := range orgPreferenceValues {
+		preferenceValueMap[preferenceValue.PreferenceId] = preferenceValue.PreferenceValue
+	}
+
+	// update in the above filtered list wherver value present in the map
+	for _, preference := range preferences {
+		isEnabledForOrgScope := preference.IsEnabledForScope(OrgAllowedScope)
+		preferenceWithValue := AllPreferences{}
+		preferenceWithValue.Key = preference.Key
+		preferenceWithValue.Name = preference.Name
+		preferenceWithValue.Description = preference.Description
+		preferenceWithValue.AllowedScopes = preference.AllowedScopes
+		preferenceWithValue.AllowedValues = preference.AllowedValues
+		preferenceWithValue.DefaultValue = preference.DefaultValue
+		preferenceWithValue.Range = preference.Range
+		preferenceWithValue.ValueType = preference.ValueType
+		preferenceWithValue.IsDiscreteValues = preference.IsDiscreteValues
+
+		if isEnabledForOrgScope {
+			ok, seen := preferenceValueMap[preference.Key]
+
+			if seen {
+				preferenceWithValue.Value = ok
+			} else {
+				preferenceWithValue.Value = preference.DefaultValue
+			}
+
+			preferenceWithValue.Value = preference.SanitizeValue(preferenceWithValue.Value)
+			allOrgPreferences = append(allOrgPreferences, preferenceWithValue)
+		}
+	}
+	return &allOrgPreferences, nil
+}
+
+// user preference functions
+func GetUserPreference(ctx context.Context, preferenceId string, orgId string) (*PreferenceKV, *model.ApiError) {
+
+	user := common.GetUserFromContext(ctx)
+	// check if the preference key exists
```

Comment:
This is an interesting choice here. The user is derived from the request context and is specific to the http entrypoint. The derivation of user in such a way makes it a part of the http entrypoint and as such the user should be passed in the GetUserPreference function as an argument.

By deriving the user from the context here, you are essentially saying, "Hey, this function can only be called if it comes from an http request."

Reply:
this is a nice thought here. makes sense to pass the user reference from the consumer here allowing this function to be reusable somewhere else as well when required.

Reply:
done